### ISSUE TITLE
Implement language files for all in-game text

### ISSUE THREAD

**Issue by ShadowApex** [Tuxemon/Tuxemon#13]
Currently all in-game menus and text are hard-coded in English. This can make it hard to implemented non-English translations of the game. To remedy this, a language file system should be implemented to make it easier for non-English translations of the game.

**Comment by sdoss6** [comment]
Would it be worthwhile to set up a database system using something like SQLite or XML to store and organize all of the text of dialogue and menu items? Having it all in one place would make it easier for someone to translate.

**Comment by bitcraft** [comment]
I agree that storing all translatable text in one place is the way to go.  Generally, I prefer JSON to XML.  For such a small project, I don't see SQLite being required.  I think the least effort in maintaining translations would be to use unique identifiers (just a string?) in the map files, with a simple key/value language file.  Maybe one file for each language, English included.  It will certainly be easier to spot spelling and grammar mistakes if we store the dialog text in a text file instead of the tiled TMX.

**Comment by ShadowApex** [comment]
We definitely should use a system to organize all the translations. Right now we're using JSON for pretty much all of our in-game data (monsters, items, encounters, etc.) so I think it'd make sense to stick with that. It's similar to XML.

My thoughts were that we'd have a directory called `resources/db/locale` that would store text files of the translations in JSON format. Just as bitcraft suggested, they'd be simple key/value identifiers that we could use in-game or during map creation to display the correct text.

It would look something like this:

`resources/db/locale/en_US.json`
```json
{
    "bag": "Bag", 
    "journal": "Journal", 
    "monster": "Monster",
    "dialog1": "Hello there young chap!"
}
```

`resources/db/locale/de_DE.json`
```json
{
    "bag": "Tasche", 
    "journal": "Zeitschrift", 
    "monster": "Kreatur",
    "dialog1": "Hallo dort junge kerl!"
}
```

**Comment by bitcraft** [comment]
Exactly what I had in mind.

**Comment by tamashihoshi** [comment]
Using a key-value system is pretty neat but there is something one should never forget: one word might have different meanings in different languages...
lets take shadowapex's example:

"journal": "Zeitschrift"
"Zeitschrift" is more like a magazine, but a journal could also be called "Journal" (yes we have that in german too, it's similar to "log") or "Tagebuch" (which is more like "diary")

It's pretty simple to solve this problem though. When creating those en_US and de_DE json files just clearify what journal we are talking about exactly. Is it an item? Then use "journal_item" as a key. Is it the menu? Then use "journal_menu".
Both keys will lead in the en_US.json to "Journal" but in the de_DE.json it might lead to "Journal" and "Zeitschrift".

I... hope that makes sense and you understand what I wanna say :D
If this suggestion is good, I'll try to work on a translation. Somewhen...

Oh by the way: Since all stuff is hard-coded, does that mean I would have to edit all main files too?

**Comment by bitcraft** [comment]
Good point about word meanings. To solve that we can make sure our key is
clear like "journal_item" and "journal_menu" or something similar.
About changing the hard coded values, I don't expect translators to deal
with the code. It would be up to somebody on the programming team to change
the English words and phrases to the keys that would be used in the
translation dictionary.

**Comment by tamashihoshi** [comment]
I have another question about translations... with the locale suggestion by shadowapex we could translate dialogues and menus easily but what should we do about the techniques? currently their name is also semi-hard-coded in english in their json file

example: pound.json
`{
    "animation": "powerslash",
    "sfx": "blaster1.ogg",
    "category": "physical",
    "effects": [
        "damage"
    ], 
    "id": 1, 
    "name": "Pound", 
    "power": 50, 
    "types": [
        "Normal"
    ]
}`

any idea how to translate those? does every attack also get a key.. like "attack_pound" or something?

I think if we solve that problem we might start with the translations

**Comment by ShadowApex** [comment]
Yes, we could use the key `technique_<name>` for technique names. Once we have some translations available we can change the technique json files to include the key name instead of the English name and load the correct translation.

**Comment by ShadowApex** [comment]
I started building a translation file in English for some of the in-game text. It can be viewed here:
https://github.com/Tuxemon/Tuxemon/blob/development/tuxemon/resources/db/locale/en_US.json

**Comment by tamashihoshi** [comment]
I've done a de_DE.json file into the same directory as the en_US.json...
I have a few questions:
1 - "exit: exit" <- where is that used? Is this for exiting a menu or from a room? I used a word which you can use in both cases but I still wanna know what I've translated :D
2 - "log off" <- like.. log out? from multiplayer?
3 - "menu journal" <- whats that? (I really need to have a look at the menus, i know..)
4 - "menu run" <- just to clarify this: is this the "run" option when in a battle?

5 - "technique_pound"
this one is more interesting: pound is a verb which describes something being hit heavily; that's what this attack does. Something like tackle maybe. But "pound" is also the money thingy they use in britain..
I have pokemon blue and I remember how they translated the attack "pound" to "pfund" which is the british money. So my question is: do we keep this awkward translation? or do we translate it properly?
that's a weird question :D but I thought I might ask you haha

6 - "item"
I bet, that's when you're in a battle and want to use the capt. device or a potion or something... I've translated it with "item" for now since I don't know about the following: How long can I make the words? I mean, the correct translation would be "Gegenstand", maybe "Inventar" from inventory.. but those are pretty long. Should I keep "Item"? We use that word in games too..

7 - ßäöü
What should I do with these? Can I use em? Or do I need to write them differently to work (like in html with &auml; for ä)?

**Comment by tamashihoshi** [comment]
Erm... I don't see the file I made in the repository.. did I do something wrong or do I have to wait for shadowapex to accept the file?

also something funny: "like in html with ä for ä" makes total sense? no? that has a simple reason: I had actually written & a uml ; and it automatically converted it into ä ._.

**Comment by ShadowApex** [comment]
> 1 - "exit: exit" <- where is that used? Is this for exiting a menu or from a room? I used a word which you can use in both cases but I still wanna know what I've translated :D

Currently that is being used for the exit menu. It might be better for me to rename that to `menu_exit` or `menu_game_quit`. Right now I'm trying to adhere to anything prefixed with `menu_` is text in a menu item.

> 2 - "log off" <- like.. log out? from multiplayer?

This is like logging off from a computer.

> 3 - "menu journal" <- whats that? (I really need to have a look at the menus, i know..)

This is the "Journal" menu item text in the main menu. Used for storing information about what the player needs to do (for example, quests, etc.).

> 4 - "menu run" <- just to clarify this: is this the "run" option when in a battle?

That's right. It's for running away from a battle.

> 5 - "technique_pound"

I think we could translate these properly.

> 6 - "item"

Yes, this is the menu item to select an item to use in battle. I think we should focus on keeping the translated text short, as it could definitely affect how it is presented in-game. You may also try abbreviating words if you can. For example, the "Tuxemon" text could alternatively be displayed as "TXMN" to keep it short.

> 7 - ßäöü
What should I do with these? Can I use em? Or do I need to write them differently to work (like in html with ä for ä)?

You can use the normal text (non-html) for these kinds of characters. One thing to keep in mind though is that certain fonts we use may not have non-English characters. Right now we're almost exclusively using the [Press Start 2P](http://www.dafont.com/press-start-2p.font) font, which supports some characters that are in German, but don't have support for characters like the ones found in Chinese or Japanese kana.

> Erm... I don't see the file I made in the repository.. did I do something wrong or do I have to wait for shadowapex to accept the file?

You won't have access to add the file you made directly to the official Tuxemon project. Normally to request a change to Tuxemon through GitHub you need to "fork" the Tuxemon project into your own copy of the project, upload your change there, then make a "Pull Request" to request that the Tuxemon project accept the changes you made in your personal repository. 

Since those steps are a lot for a new user, you could alternatively upload your translations using [GitHub Gist](https://gist.github.com/) or using a site like [Hastebin](http://hastebin.com/) and post a comment with a link to it here.

**Comment by tamashihoshi** [comment]
That forking and pulling around confuses me :( I need more time to understand that..
It even took me quite a while to understand that the file was in the development branch and now the main branch..
anyways, here are the translations: https://gist.github.com/tamashihoshi/9bab5d807ec1db1fedad

About the font: Do you think we should look for one font combining all chars? Or are seperate fonts also okay? Sounds a little weird but I don't think japanese hiragana/katagana/kanji stuff looks similar to our letters so it should be fine. Maybe?

Something different about the translations: Does it make any sense to write a version into the json files? That could give us information on how far the translation into a language is.. I mean if you edit the en_US.json the de_DE.json will not translate by itself. A version check could be used to display some kind of "Some Words are not translated in this language yet. Continue?" message if you select the language ingame. Maybe.

PS: I'm not quite happy with the translation to "capture device".. it sounds weird in german. Should we maybe change the item name? Until then I'm thinking of a better translation...

**Comment by tamashihoshi** [comment]
German translations: #97

What's left to do: Implement a variable somewhere which saves the language aka the path to the json file and get stuff out of it..

does it make sense to read the strings out of the json file when needed? or should we parse the file once at the very beginning into a list with a simple json.loads() and keep the list (sounds weird)?

Does json.loads even make sense for us or is there a better way somehow? I'm still a pynoob ._.

**Comment by bitcraft** [comment]
Congratz on the first pull request !  Just my comments on it; the translation functions haven't been planned out, but the way I'd do it is like the following (I'm showing the API, not implementation):

```python
from tuxemon.core.intl import Translator

tr = Translator('de_DE')

print(tr('bag'))
"Tasche"

print(tr('dialog1'))
"Kreatur"

print(tr('not_implemented'))
"Diese Funktion ist noch nicht implementiert."
```

As far as implementation goes, the best way, imo, would be to create a Translator object, which loads the translation into memory, then keep one copy of it.  It could be a singleton or even just attach it to the 'Game' object.  If eventually the text gets to be too large to hold in memory (not likely), then we can devise a system for storing on disk.

**Comment by tamashihoshi** [comment]
thanks! :D

the translator object sounds good...
"If eventually the text gets to be too large to hold in memory (not likely)" <- I was taught to ignore "not likely" and always go the safe way but I hope you know what you're doing. I bet you have #58 in mind, so I think you know what you're doing :D

I think there is one last thing needed then; the en_US and de_DE stuff should only contain strings, that are needed for the basic game itself. Stuff like dialogues inside the game should be seperated.. I remember something about different stories or so you could choose from.. so if we plan on doing stuff like downloadable levels, the levels/maps/stories should contain their own item/story strings inside. So maybe an advanced string-json system is needed...
If I'm mixing up something then nevermind that one xD

**Comment by ShadowApex** [comment]
Great work on your first pull request! Everything looks perfect!

> the translator object sounds good...
"If eventually the text gets to be too large to hold in memory (not likely)" <- I was taught to ignore "not likely" and always go the safe way but I hope you know what you're doing. I bet you have #58 in mind, so I think you know what you're doing :D

Using a translator object will actually let us do this if we need to in the future. If we get to the point where we need to load strings from disk we can extend and modify the translator object to do that. It's actually highly unlikely we'll ever actually need to do that though. As an example, the entire 1,500 page novel "War and Peace" (which has over 580,000 words) is only [1.7 MB in size](https://www.goodreads.com/ebooks/download/656.War_and_Peace).

> I think there is one last thing needed then; the en_US and de_DE stuff should only contain strings, that are needed for the basic game itself. Stuff like dialogues inside the game should be seperated.. I remember something about different stories or so you could choose from.. so if we plan on doing stuff like downloadable levels, the levels/maps/stories should contain their own item/story strings inside. So maybe an advanced string-json system is needed...

All downloadable campaigns in the future will actually have their own `resources/` folder that will contain all the sprites, music, maps, and locale data separate from the main campaign. I think having a single json file for the main campaign is fine for now, but I'll definitely consider splitting it up if it gets too hard to work with. Thanks for the suggestion!

**Comment by tamashihoshi** [comment]
Thank you very much ShadowApex!
I think I will try to work on a translator object soon... unless somebody else wants to of course :D
the translation isn't that mega-much of a priority now I think.. so I can slooooooowly take care of this while you can focus on more important things :P

**Comment by tamashihoshi** [comment]
if you have time, take a tiny little look at this guy:
https://github.com/tamashihoshi/Tuxemon/blob/development/tuxemon/core/intl.py

I've put some questions/ideas as comments after some lines....

a quick explanation:
- the first part is just copy-pasted from the main tuxemon.py or something; no changes made here
- init: create translator object with lang (en_US, de_DE)
idea: 2 "different" translator objects for campaign/main translation. main only translates strings that are always needed for anything (i.e. "YOU WON!") and campaign has all other stuff (dialogues etc)
- trans: returns key from the initally loaded json file...


for our current state this should already work but I haven't tested it yet (that's why I didnt do a PR)
needs to be updated of course, when the downloadable campaigns are implemented..

usage:
import PathToFile/Translator
tr = Translator('de_DE')
tr.trans('combat_victory')
(returns: 'DU GEWINNST!')

**Comment by Clonewayx** [comment]
Have you thought about using some translating tool?
For example https://translatewiki.net

**Comment by ShadowApex** [comment]
In-game translation has been implemented in the latest commit!: 495cf6b20ef74e46aa512a8d91c060dbcb495bac

Much of the `Translator` object was based off of [tamashihoshi](https://github.com/tamashihoshi)'s example that he provided. You can check out the new translator object here:
https://github.com/Tuxemon/Tuxemon/blob/development/tuxemon/core/components/locale.py

Right now this is how in-game translation is currently handled:
* Game language is determined via the `locale` config option under the `[game]` section in `~/.tuxemon/tuxemon.cfg`.
* If no locale is set in `tuxemon.cfg`, "en_US" is selected by default.
* If a translation does not exist for a certain word/phrase in a given language, translation will fall back to "en_US". If a translation doesn't exist in "en_US" either, "Locale Error" text will be displayed instead.
* All items and techniques have a `name_trans` and `description_trans` property that will be used for translation lookup. The `name` and `description` properties will be loaded as a fall back.

Here are some current issues:
* I needed to add a few more entries to "en_US" to finish replacing all in-game text. Other languages will need to be updated to include these new entries.
* Some menu item text in certain languages is larger than the menu width. We'll need to find a way to either increase the size of the menu or use abbreviated words (e.g. "Tuxemon" could be "TXMN").
